blocks/api/categories.js

```diff
@@ -20,10 +23,111 @@ const categories = [
 ];
 
 /**
- * Returns all the block categories
+ * Returns all the block categories.
  *
- * @return {Array} Block categories
+ * @return {Array} categories
  */
 export function getCategories() {
 	return categories;
 }
+
+/**
+ * Register a new block Category.
+ *
+ * @param {Array} category e.g {slug: 'custom', title: __('Custom Blocks')}
+ *
+ * @return {Array} categories
+ */
+export function registerCategory( category ) {
+	if ( ! category ) {
+		console.error(
+			'The block Category must be defined'
+		);
+		return;
+	}
+	if ( ! category.slug ) {
+		console.error(
+			'The block Category slug must be defined'
+		);
+		return;
+	}
+	if ( ! /^[a-z0-9-]+$/.test( category.slug ) ) {
```

Comment:
If we're going to define a rule for what category names can be it could be nice to make it explicit somewhere where others can find it without having to dive deep into the code: maybe a constant in the top of the module scope?

```js
/**
 * @type {RegExp}
 * @const
 *
 * Category names must be a combination of lower-case letters, numbers, and hypens
 */
const categoryNamePattern = /^[a-z0-9-]+$/

// or…
//     ^[a-z]   - must start with a letter
//     (?: … )+ - any number of the following (non-capture because it's all part of a whole name)
//         -*       - any number of hyphens
//         [a-z0-9] - followed by a letter or number
const categoryNamePattern = /^[a-z](?:-*[a-z0-9])+$/
```